execution-engine/src/hcall/wasmi.rs

```diff
@@ -10,38 +10,69 @@
     RuntimeArgs, RuntimeValue, Signature, TableDescriptor, TableRef, Trap, ValueType,
+    HostError, TrapKind
 };
 
-use crate::{
-    error::{
-        common::VeracruzError,
-        wasmi::{mk_error_code, mk_host_trap},
-    },
-    hcall::common::{
-        sha_256_digest, ExecutionEngine, DataSourceMetadata, EntrySignature, FatalHostError, HCallError,
-        HostProvisioningError, HostProvisioningState, LifecycleState, HCALL_GETRANDOM_NAME,
+use crate::hcall::{
+    common::{
+        ExecutionEngine, EntrySignature, EngineReturnCode,
+        HostProvisioningError, FatalEngineError, VFSService, HCALL_GETRANDOM_NAME,
         HCALL_HAS_PREVIOUS_RESULT_NAME, HCALL_INPUT_COUNT_NAME, HCALL_INPUT_SIZE_NAME,
         HCALL_PREVIOUS_RESULT_SIZE_NAME, HCALL_READ_INPUT_NAME, HCALL_READ_PREVIOUS_RESULT_NAME,
         HCALL_READ_STREAM_NAME, HCALL_STREAM_COUNT_NAME, HCALL_STREAM_SIZE_NAME,
         HCALL_WRITE_OUTPUT_NAME,
     },
 };
+use veracruz_utils::policy::principal::Principal;
+#[cfg(any(feature = "std", feature = "tz", feature = "nitro"))]
+use std::sync::{Mutex, Arc};
+#[cfg(feature = "sgx")]
+use std::sync::{SgxMutex as Mutex, Arc};
+use crate::hcall::buffer::VFS;
 
 ////////////////////////////////////////////////////////////////////////////////
 // The WASMI host provisioning state.
 ////////////////////////////////////////////////////////////////////////////////
 
+#[typetag::serde]
+impl HostError for FatalEngineError {}
+
+/// The return type for H-Call implementations.
+///
+/// From *the viewpoint of the host* a H-call can either fail spectacularly
+/// with a runtime trap, in which case `Err(err)` is returned, with `err`
+/// detailing what went wrong, and the Veracruz host thereafter terminating
+/// or otherwise entering an error state, or succeeds with `Ok(())`.
+///
+/// From *the viewpoint of the WASM program* a H-call can either fail
+/// spectacularly, as above, in which case WASM program execution is aborted
+/// with the WASM program itself not being able to do anything about this,
+/// succeeds with the desired effect and a success error code returned, or
+/// fails with a recoverable error in which case the error code details what
+/// went wrong and what can be done to fix it.
+pub(crate) type HCallError = Result<EngineReturnCode, FatalEngineError>;
+
 /// The WASMI host provisioning state: the `HostProvisioningState` with the
 /// Module and Memory type-variables specialised to WASMI's `ModuleRef` and
 /// `MemoryRef` type.
-pub(crate) type WasmiHostProvisioningState = HostProvisioningState<ModuleRef, MemoryRef>;
+pub(crate) struct WasmiHostProvisioningState {
+    vfs : VFSService,
+    /// A reference to the WASM program module that will actually execute on
+    /// the input data sources.
+    program_module: Option<ModuleRef>,
+    /// A reference to the WASM program's linear memory (or "heap").
+    memory: Option<MemoryRef>,
+    /// Ref to the program that is executed
+    program: Principal,
```

Comment:
What is the intuition, here?  Why is a program a `Principal`?

Reply:
I probably should use a better name, principal includes participant (client) and program. The capability is bound to a principal. For example, if a client and a program want to read or write a file, it will first check the capability.